File: matlab/sortingalgos/bubbleSort.m
function inArray = bubbleSort(inArray)

for i = 1:numel(inArray) - 1
   numSwaps = 0;
    for j = 1:numel(inArray) - i %optimization: last i elements are already sorted.
        if(inArray(j) > inArray(j+1))
            temp = inArray(j);
            inArray(j) = inArray(j+1);
            inArray(j+1) = temp;
            numSwaps = numSwaps + 1;
        end 
    end
    if(numSwaps == 0) 
        % this means that no swap has happened in the ith iteration
        % and hence means that the list is already sorted.
        break;
    end
end

end

File: python/sortingAlgos/bubbleSort.py
def bubbleSort(arrayIn):
    for i in range(1, len(arrayIn)):
        numSwap = 0
        for j in range(1, len(arrayIn) - i + 1): 
            #optimized: loops only until len(array) - i + 1: the last i elements are known to be sorted by now.
            if arrayIn[j-1] > arrayIn[j]:
                # the next line is a swap in python.
                arrayIn[j-1], arrayIn[j] = arrayIn[j], arrayIn[j-1]
                numSwap += 1
        if numSwap == 0:
            break # as no swaps mean that the list is already sorted.
    return arrayIn            
        
arrayIn = [2, 1, 5, 3, 6, 4]
print arrayIn
print bubbleSort(arrayIn)

print bubbleSort([1])

print bubbleSort([])
